// @abstract-foundation/agw-client/dist/esm/eip5792.js
var CallStatus;
(function (CallStatus) {
    CallStatus[CallStatus["Pending"] = 100] = "Pending";
    CallStatus[CallStatus["Confirmed"] = 200] = "Confirmed";
    CallStatus[CallStatus["OffchainFailure"] = 400] = "OffchainFailure";
    CallStatus[CallStatus["Reverted"] = 500] = "Reverted";
    CallStatus[CallStatus["PartiallyReverted"] = 600] = "PartiallyReverted";
})(CallStatus || (CallStatus = {}));
export const agwCapabilitiesV2 = {
    '0xab5': {
        atomicBatch: {
            supported: true,
        },
        atomic: {
            status: 'supported',
        },
    },
    '0x2b74': {
        atomicBatch: {
            supported: true,
        },
        atomic: {
            status: 'supported',
        },
    },
};
export function getReceiptStatus(receipt) {
    switch (receipt?.status) {
        case undefined:
            return CallStatus.Pending;
        case '0x1':
            return CallStatus.Confirmed;
        case '0x0':
            return CallStatus.Reverted;
    }
}
